app/server.js

```diff
@@ -1,13 +1,184 @@
+router.get('/products', (req, res) => {
+  res.writeHead(200, { "Content-Type": "application/json" });
+  return res.end(JSON.stringify(products));
+})
+
+// 3. Requests for a specific product (passed):
+router.get('/products/:categoryId', (req, res) => {
+  const productId = req.params.categoryId;
+  console.log(productId);
+  const product = products.filter(p => p.categoryId === productId);
+  res.writeHead(200, { "Content-Type": "application/json" });
+  return res.end(JSON.stringify(product))
+})
+
+// 4. Handle POST login request from user:
+router.post('/login', (req, res) => {
+  if(req.body.username && req.body.password){
+    let currentUser = users.find((user) => {
+        user.login.username == req.body.username &&
+        user.login.password == req.body.password  
+    });
+
+  if (currentUser) {
+    res.writeHead(200, { "Content-Type": "application/json" });
+
+    // Check if token exists:
+    let currentAccessToken = accessTokens.find((tokenObject) => 
+      tokenObject.username === currentUser.login.username
+    );
+    
+    // Update latest value to get another time period:
+    if (currentAccessToken){
+      currentAccessToken.lastUpdate = new Date();
+      return res.end(JSON.stringify(currentAccessToken.token));
+    } else {
+      // Create a new token with the user value and a 'random token'
+      let newAccessToken = {
+        username:  currentUser.login.username,
+        lastUpdated: new Date(),
+        token: uid(16)
+      }
+      accessTokens.push(newAccessToken);
+      return res.end(JSON.stringify(newAccessToken.token));
+      }
+
+    } else {
+      res.writeHead(401, 'Invalid username or password');
+      return res.end();
+    }
+  } else {
+      res.writeHead(400, 'Incorrectly formatted response');
+      return res.end();
+  }
+});
+
+// Helper method to verify tokens
+var verifyAccessToken = function(request) {
+  var parsedUrl = require('url').parse(request.url, true);
+  if (parsedUrl.query.accessToken) {
+    // Verify the access token to make sure it's valid and not expired
+    let currentAccessToken = accessTokens.find(
+      (accessToken) => 
+      accessToken.token == parsedUrl.query.accessToken &&
+      new Date() - accessToken.lastUpdated < TOKEN_VALIDITY_TIMEOUT
+    );
+
+    if (currentAccessToken) {
+      return currentAccessToken;
+    } else {
+      return null;
+    }
+
+  } else {
+    return null;
+  }
+};
+
+// Get user's cart
+router.get('/cart', function (req, res) {
+  let currentAccessToken = verifyAccessToken (req);
+
+  if(currentAccessToken) {
+    const currentUser = getUser(currentAccessToken);
+    res.writeHead(200, {'Content-Type': 'application/json'});
+    return res.end(JSON.stringify(currentUser.cart));
+  }
+    res.writeHead(401, 'Please login first')
+    return res.end();
+});
+
+// Post item to user's cart
+router.post('/cart/:id', async (req, res) => {
+  let accessToken = getValidAccessTokenFromRequest(req);
```

Comment:
This method doesn't exist anywhere in the app, so this endpoint would likely fail.